ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -0,0 +1,742 @@
+import           Ouroboros.Storage.ChainDB.Impl.LgrDB (LgrDB)
+import qualified Ouroboros.Storage.ChainDB.Impl.LgrDB as LgrDB
+import qualified Ouroboros.Storage.ChainDB.Impl.Query as Query
+import qualified Ouroboros.Storage.ChainDB.Impl.Reader as Reader
+import           Ouroboros.Storage.ChainDB.Impl.Types
+import           Ouroboros.Storage.ChainDB.Impl.VolDB (VolDB)
+import qualified Ouroboros.Storage.ChainDB.Impl.VolDB as VolDB
+
+
+-- | Perform the initial chain selection based on the tip of the ImmutableDB
+-- and the contents of the VolatileDB.
+--
+-- Returns the chosen validated chain and corresponding ledger.
+--
+-- See "## Initialization" in ChainDB.md.
+initialChainSelection
+  :: forall m blk.
+     ( MonadCatch m
+     , MonadSTM   m
+     , ProtocolLedgerView blk
+     )
+  => ImmDB m blk
+  -> VolDB m blk
+  -> LgrDB m blk
+  -> Tracer m (TraceEvent blk)
+  -> NodeConfig (BlockProtocol blk)
+  -> TVar m (Set (Point blk)) -- ^ 'cdbInvalid'
+  -> m (ChainAndLedger blk)
+initialChainSelection immDB volDB lgrDB tracer cfg invalidPoints = do
+    -- We follow the steps from section "## Initialization" in ChainDB.md
+
+    i :: Point blk <- ImmDB.getPointAtTip immDB
+    (succsOf, ledger) <- atomically $
+      (,) <$> VolDB.getSuccessors volDB
+          <*> LgrDB.getCurrent    lgrDB
+    chains <- constructChains i succsOf
+
+    -- We use the empty fragment anchored at @i@ as the current chain (and
+    -- ledger) and the default in case there is no better candidate.
+    let curChainAndLedger = mkChainAndLedger (Empty (castPoint i)) ledger
+
+    case NE.nonEmpty chains of
+      -- If there are no candidates, no chain selection is needed
+      Nothing      -> return curChainAndLedger
+      Just chains' -> fromMaybe curChainAndLedger <$>
+        chainSelection' curChainAndLedger chains'
+  where
+    -- | Use the VolatileDB to construct all chains starting from the tip of
+    -- the ImmutableDB.
+    constructChains :: Point blk -- ^ Tip of the ImmutableDB, @i@
+                    -> (Maybe (HeaderHash blk) -> Set (HeaderHash blk))
+                    -> m [AnchoredFragment (Header blk)]
+    constructChains i succsOf = flip evalStateT Map.empty $
+        mapM constructChain suffixesAfterI
+      where
+        suffixesAfterI :: [NonEmpty (HeaderHash blk)]
+        suffixesAfterI = VolDB.candidates succsOf i
+
+        constructChain :: NonEmpty (HeaderHash blk)
+                       -> StateT (Map (HeaderHash blk) (Header blk))
+                                 m
+                                 (AnchoredFragment (Header blk))
+        constructChain hashes =
+          AF.fromOldestFirst (castPoint i) <$>
+          mapM (getKnownHeaderThroughCache volDB) (NE.toList hashes)
+
+    -- | Perform chain selection (including validation) on the given
+    -- candidates.
+    --
+    -- PRECONDITION: all candidates are anchored at @i@.
+    chainSelection' :: HasCallStack
```

Comment:
Check that each candidate is preferred over the current chain, which is a precondition of `chainSelection`.